Plugin support for visual shaders

Chaosus opened the pull request:
This PR allows user to create custom nodes for visual shader editor. ~~The creation process is similar to standart EditorPlugin (described in http://docs.godotengine.org/en/latest/tutorials/plugins/editor/making_plugins.html).~~

Fix #3349

Here is example of creation Noise function plugin (using shader code from https://github.com/curly-brace/Godot-3.0-Noise-Shaders)

```
tool
extends VisualShaderNodeCustom
class_name VisualShaderNodeNoise

func _get_name():
	return "Noise"

func _get_category():
	return "CoolLib"

func _get_subcategory():
	return "Effects"

func _get_description():
	return "My favorite noise function"

func _get_return_icon_type():
	return 0 # scalar

func _get_input_port_count():
	return 2
	
func _get_input_port_type(port):
	match port:
		0:
			return 1 # vector
		1:
			return 0 # scalar
	return 0
	
func _get_input_port_name(port):
	match port:
		0:
			return "offset"
		1:
			return "scale"
	return ""

func _get_output_port_count():
	return 1

func _get_output_port_type(port):
	return 0 # scalar

func _get_output_port_name(port):
	return "output"

func _get_global_code(mode):
	var code := ""
	code += """
	vec3 mod289_3(vec3 x) {
	    return x - floor(x * (1.0 / 289.0)) * 289.0;
	}
	
	vec4 mod289_4(vec4 x) {
	    return x - floor(x * (1.0 / 289.0)) * 289.0;
	}
	
	vec4 permute(vec4 x) {
	    return mod289_4(((x*34.0)+1.0)*x);
	}
	
	vec4 taylorInvSqrt(vec4 r) {
	    return 1.79284291400159 - 0.85373472095314 * r;
	}
	
	vec3 fade(vec3 t) {
	    return t * t * t * (t * (t * 6.0 - 15.0) + 10.0);
	}
	
	// Classic Perlin noise
	float cnoise(vec3 P) {
	    vec3 Pi0 = floor(P); // Integer part for indexing
	    vec3 Pi1 = Pi0 + vec3(1.0); // Integer part + 1
	    Pi0 = mod289_3(Pi0);
	    Pi1 = mod289_3(Pi1);
	    vec3 Pf0 = fract(P); // Fractional part for interpolation
	    vec3 Pf1 = Pf0 - vec3(1.0); // Fractional part - 1.0
	    vec4 ix = vec4(Pi0.x, Pi1.x, Pi0.x, Pi1.x);
	    vec4 iy = vec4(Pi0.yy, Pi1.yy);
	    vec4 iz0 = vec4(Pi0.z);
	    vec4 iz1 = vec4(Pi1.z);
	    
	    vec4 ixy = permute(permute(ix) + iy);
	    vec4 ixy0 = permute(ixy + iz0);
	    vec4 ixy1 = permute(ixy + iz1);
	    
	    vec4 gx0 = ixy0 * (1.0 / 7.0);
	    vec4 gy0 = fract(floor(gx0) * (1.0 / 7.0)) - 0.5;
	    gx0 = fract(gx0);
	    vec4 gz0 = vec4(0.5) - abs(gx0) - abs(gy0);
	    vec4 sz0 = step(gz0, vec4(0.0));
	    gx0 -= sz0 * (step(0.0, gx0) - 0.5);
	    gy0 -= sz0 * (step(0.0, gy0) - 0.5);
	    
	    vec4 gx1 = ixy1 * (1.0 / 7.0);
	    vec4 gy1 = fract(floor(gx1) * (1.0 / 7.0)) - 0.5;
	    gx1 = fract(gx1);
	    vec4 gz1 = vec4(0.5) - abs(gx1) - abs(gy1);
	    vec4 sz1 = step(gz1, vec4(0.0));
	    gx1 -= sz1 * (step(0.0, gx1) - 0.5);
	    gy1 -= sz1 * (step(0.0, gy1) - 0.5);
	    
	    vec3 g000 = vec3(gx0.x,gy0.x,gz0.x);
	    vec3 g100 = vec3(gx0.y,gy0.y,gz0.y);
	    vec3 g010 = vec3(gx0.z,gy0.z,gz0.z);
	    vec3 g110 = vec3(gx0.w,gy0.w,gz0.w);
	    vec3 g001 = vec3(gx1.x,gy1.x,gz1.x);
	    vec3 g101 = vec3(gx1.y,gy1.y,gz1.y);
	    vec3 g011 = vec3(gx1.z,gy1.z,gz1.z);
	    vec3 g111 = vec3(gx1.w,gy1.w,gz1.w);
	    
	    vec4 norm0 = taylorInvSqrt(vec4(dot(g000, g000), dot(g010, g010), dot(g100, g100), dot(g110, g110)));
	    g000 *= norm0.x;
	    g010 *= norm0.y;
	    g100 *= norm0.z;
	    g110 *= norm0.w;
	    vec4 norm1 = taylorInvSqrt(vec4(dot(g001, g001), dot(g011, g011), dot(g101, g101), dot(g111, g111)));
	    g001 *= norm1.x;
	    g011 *= norm1.y;
	    g101 *= norm1.z;
	    g111 *= norm1.w;
	    
	    float n000 = dot(g000, Pf0);
	    float n100 = dot(g100, vec3(Pf1.x, Pf0.yz));
	    float n010 = dot(g010, vec3(Pf0.x, Pf1.y, Pf0.z));
	    float n110 = dot(g110, vec3(Pf1.xy, Pf0.z));
	    float n001 = dot(g001, vec3(Pf0.xy, Pf1.z));
	    float n101 = dot(g101, vec3(Pf1.x, Pf0.y, Pf1.z));
	    float n011 = dot(g011, vec3(Pf0.x, Pf1.yz));
	    float n111 = dot(g111, Pf1);
	    
	    vec3 fade_xyz = fade(Pf0);
	    vec4 n_z = mix(vec4(n000, n100, n010, n110), vec4(n001, n101, n011, n111), fade_xyz.z);
	    vec2 n_yz = mix(n_z.xy, n_z.zw, fade_xyz.y);
	    float n_xyz = mix(n_yz.x, n_yz.y, fade_xyz.x); 
	    return 2.2 * n_xyz;
	}
	
	// Classic Perlin noise, periodic variant
	float pnoise(vec3 P, vec3 rep) {
	    vec3 Pi0 = mod(floor(P), rep); // Integer part, modulo period
	    vec3 Pi1 = mod(Pi0 + vec3(1.0), rep); // Integer part + 1, mod period
	    Pi0 = mod289_3(Pi0);
	    Pi1 = mod289_3(Pi1);
	    vec3 Pf0 = fract(P); // Fractional part for interpolation
	    vec3 Pf1 = Pf0 - vec3(1.0); // Fractional part - 1.0
	    vec4 ix = vec4(Pi0.x, Pi1.x, Pi0.x, Pi1.x);
	    vec4 iy = vec4(Pi0.yy, Pi1.yy);
	    vec4 iz0 = vec4(Pi0.z);
	    vec4 iz1 = vec4(Pi1.z);
	    
	    vec4 ixy = permute(permute(ix) + iy);
	    vec4 ixy0 = permute(ixy + iz0);
	    vec4 ixy1 = permute(ixy + iz1);
	    
	    vec4 gx0 = ixy0 * (1.0 / 7.0);
	    vec4 gy0 = fract(floor(gx0) * (1.0 / 7.0)) - 0.5;
	    gx0 = fract(gx0);
	    vec4 gz0 = vec4(0.5) - abs(gx0) - abs(gy0);
	    vec4 sz0 = step(gz0, vec4(0.0));
	    gx0 -= sz0 * (step(0.0, gx0) - 0.5);
	    gy0 -= sz0 * (step(0.0, gy0) - 0.5);
	    
	    vec4 gx1 = ixy1 * (1.0 / 7.0);
	    vec4 gy1 = fract(floor(gx1) * (1.0 / 7.0)) - 0.5;
	    gx1 = fract(gx1);
	    vec4 gz1 = vec4(0.5) - abs(gx1) - abs(gy1);
	    vec4 sz1 = step(gz1, vec4(0.0));
	    gx1 -= sz1 * (step(0.0, gx1) - 0.5);
	    gy1 -= sz1 * (step(0.0, gy1) - 0.5);
	    
	    vec3 g000 = vec3(gx0.x,gy0.x,gz0.x);
	    vec3 g100 = vec3(gx0.y,gy0.y,gz0.y);
	    vec3 g010 = vec3(gx0.z,gy0.z,gz0.z);
	    vec3 g110 = vec3(gx0.w,gy0.w,gz0.w);
	    vec3 g001 = vec3(gx1.x,gy1.x,gz1.x);
	    vec3 g101 = vec3(gx1.y,gy1.y,gz1.y);
	    vec3 g011 = vec3(gx1.z,gy1.z,gz1.z);
	    vec3 g111 = vec3(gx1.w,gy1.w,gz1.w);
	    
	    vec4 norm0 = taylorInvSqrt(vec4(dot(g000, g000), dot(g010, g010), dot(g100, g100), dot(g110, g110)));
	    g000 *= norm0.x;
	    g010 *= norm0.y;
	    g100 *= norm0.z;
	    g110 *= norm0.w;
	    vec4 norm1 = taylorInvSqrt(vec4(dot(g001, g001), dot(g011, g011), dot(g101, g101), dot(g111, g111)));
	    g001 *= norm1.x;
	    g011 *= norm1.y;
	    g101 *= norm1.z;
	    g111 *= norm1.w;
	    
	    float n000 = dot(g000, Pf0);
	    float n100 = dot(g100, vec3(Pf1.x, Pf0.yz));
	    float n010 = dot(g010, vec3(Pf0.x, Pf1.y, Pf0.z));
	    float n110 = dot(g110, vec3(Pf1.xy, Pf0.z));
	    float n001 = dot(g001, vec3(Pf0.xy, Pf1.z));
	    float n101 = dot(g101, vec3(Pf1.x, Pf0.y, Pf1.z));
	    float n011 = dot(g011, vec3(Pf0.x, Pf1.yz));
	    float n111 = dot(g111, Pf1);
	    
	    vec3 fade_xyz = fade(Pf0);
	    vec4 n_z = mix(vec4(n000, n100, n010, n110), vec4(n001, n101, n011, n111), fade_xyz.z);
	    vec2 n_yz = mix(n_z.xy, n_z.zw, fade_xyz.y);
	    float n_xyz = mix(n_yz.x, n_yz.y, fade_xyz.x); 
	    return 2.2 * n_xyz;
	}"""
	return code

func _get_code(p_input_vars, p_output_vars, mode, type):
	var code = ""
	code += p_output_vars[0] + " = cnoise(vec3((UV + " + p_input_vars[0] + ".xy) * " + p_input_vars[1] + ", TIME)) * 0.5 + 0.5;";
	return code
```

~~After enabling the plugin in editor settings, your visual shader editor member tree will contains a new item:~~

After reloading editor, your visual shader editor member tree will contains a new item:

![image](https://user-images.githubusercontent.com/3036176/62821037-98774880-bb76-11e9-9157-febca28c1c85.png)

And the final looking function in the graph looks like this:

![image](https://user-images.githubusercontent.com/3036176/62821089-e2acf980-bb77-11e9-8b0c-0315cb3a5c60.png)

PS: If you see error message "editor\editor_settings.cpp:1219 - Condition ' err != OK ' is true." is not related to this PR, its just an already existed bug.

Chaosus committed:
Plugin support for visual shaders

clayjohn commented:
Looks good to me! The design is very elegant and should allow us to greatly expand the VisualShader without adding bulk to the engine. Next step is making a demo project with dozens of useful nodes ;)

reduz commented:
This looks really good, although I think we should start using the class_name approach for these things (instead of manually registering), so editors simply check the inheritance for those classes and use them(for this and other plugin types) since they are available in the editor.

Chaosus commented:
ok, marks PR WIP for now, will waste some time for research

Chaosus commented:
Updated as @reduz requested !

Chaosus commented:
Example above can be simply writted as

tool
extends VisualShaderNodeCustom
class_name VisualShaderNodeNoise

func _get_name():
	return "Noise"

func _get_category():
	return "CoolLib"

func _get_sub_category():
	return "Effects"

func _get_description():
	return "My favorite noise function"

func _get_return_icon_type():
	return 0 # scalar

func _get_input_port_count():
	return 2
	
func _get_input_port_type(port):
	match port:
		0:
			return 1 # vector
		1:
			return 0 # scalar
	return 0
	
func _get_input_port_name(port):
	match port:
		0:
			return "offset"
		1:
			return "scale"
	return ""

func _get_output_port_count():
	return 1

func _get_output_port_type(port):
	return 0 # scalar

func _get_output_port_name(port):
	return "output"

func _get_global_code():
	var code := ""
	code += """
	vec3 mod289_3(vec3 x) {
	    return x - floor(x * (1.0 / 289.0)) * 289.0;
	}
	
	vec4 mod289_4(vec4 x) {
	    return x - floor(x * (1.0 / 289.0)) * 289.0;
	}
	
	vec4 permute(vec4 x) {
	    return mod289_4(((x*34.0)+1.0)*x);
	}
	
	vec4 taylorInvSqrt(vec4 r) {
	    return 1.79284291400159 - 0.85373472095314 * r;
	}
	
	vec3 fade(vec3 t) {
	    return t * t * t * (t * (t * 6.0 - 15.0) + 10.0);
	}
	
	// Classic Perlin noise
	float cnoise(vec3 P) {
	    vec3 Pi0 = floor(P); // Integer part for indexing
	    vec3 Pi1 = Pi0 + vec3(1.0); // Integer part + 1
	    Pi0 = mod289_3(Pi0);
	    Pi1 = mod289_3(Pi1);
	    vec3 Pf0 = fract(P); // Fractional part for interpolation
	    vec3 Pf1 = Pf0 - vec3(1.0); // Fractional part - 1.0
	    vec4 ix = vec4(Pi0.x, Pi1.x, Pi0.x, Pi1.x);
	    vec4 iy = vec4(Pi0.yy, Pi1.yy);
	    vec4 iz0 = vec4(Pi0.z);
	    vec4 iz1 = vec4(Pi1.z);
	    
	    vec4 ixy = permute(permute(ix) + iy);
	    vec4 ixy0 = permute(ixy + iz0);
	    vec4 ixy1 = permute(ixy + iz1);
	    
	    vec4 gx0 = ixy0 * (1.0 / 7.0);
	    vec4 gy0 = fract(floor(gx0) * (1.0 / 7.0)) - 0.5;
	    gx0 = fract(gx0);
	    vec4 gz0 = vec4(0.5) - abs(gx0) - abs(gy0);
	    vec4 sz0 = step(gz0, vec4(0.0));
	    gx0 -= sz0 * (step(0.0, gx0) - 0.5);
	    gy0 -= sz0 * (step(0.0, gy0) - 0.5);
	    
	    vec4 gx1 = ixy1 * (1.0 / 7.0);
	    vec4 gy1 = fract(floor(gx1) * (1.0 / 7.0)) - 0.5;
	    gx1 = fract(gx1);
	    vec4 gz1 = vec4(0.5) - abs(gx1) - abs(gy1);
	    vec4 sz1 = step(gz1, vec4(0.0));
	    gx1 -= sz1 * (step(0.0, gx1) - 0.5);
	    gy1 -= sz1 * (step(0.0, gy1) - 0.5);
	    
	    vec3 g000 = vec3(gx0.x,gy0.x,gz0.x);
	    vec3 g100 = vec3(gx0.y,gy0.y,gz0.y);
	    vec3 g010 = vec3(gx0.z,gy0.z,gz0.z);
	    vec3 g110 = vec3(gx0.w,gy0.w,gz0.w);
	    vec3 g001 = vec3(gx1.x,gy1.x,gz1.x);
	    vec3 g101 = vec3(gx1.y,gy1.y,gz1.y);
	    vec3 g011 = vec3(gx1.z,gy1.z,gz1.z);
	    vec3 g111 = vec3(gx1.w,gy1.w,gz1.w);
	    
	    vec4 norm0 = taylorInvSqrt(vec4(dot(g000, g000), dot(g010, g010), dot(g100, g100), dot(g110, g110)));
	    g000 *= norm0.x;
	    g010 *= norm0.y;
	    g100 *= norm0.z;
	    g110 *= norm0.w;
	    vec4 norm1 = taylorInvSqrt(vec4(dot(g001, g001), dot(g011, g011), dot(g101, g101), dot(g111, g111)));
	    g001 *= norm1.x;
	    g011 *= norm1.y;
	    g101 *= norm1.z;
	    g111 *= norm1.w;
	    
	    float n000 = dot(g000, Pf0);
	    float n100 = dot(g100, vec3(Pf1.x, Pf0.yz));
	    float n010 = dot(g010, vec3(Pf0.x, Pf1.y, Pf0.z));
	    float n110 = dot(g110, vec3(Pf1.xy, Pf0.z));
	    float n001 = dot(g001, vec3(Pf0.xy, Pf1.z));
	    float n101 = dot(g101, vec3(Pf1.x, Pf0.y, Pf1.z));
	    float n011 = dot(g011, vec3(Pf0.x, Pf1.yz));
	    float n111 = dot(g111, Pf1);
	    
	    vec3 fade_xyz = fade(Pf0);
	    vec4 n_z = mix(vec4(n000, n100, n010, n110), vec4(n001, n101, n011, n111), fade_xyz.z);
	    vec2 n_yz = mix(n_z.xy, n_z.zw, fade_xyz.y);
	    float n_xyz = mix(n_yz.x, n_yz.y, fade_xyz.x); 
	    return 2.2 * n_xyz;
	}
	
	// Classic Perlin noise, periodic variant
	float pnoise(vec3 P, vec3 rep) {
	    vec3 Pi0 = mod(floor(P), rep); // Integer part, modulo period
	    vec3 Pi1 = mod(Pi0 + vec3(1.0), rep); // Integer part + 1, mod period
	    Pi0 = mod289_3(Pi0);
	    Pi1 = mod289_3(Pi1);
	    vec3 Pf0 = fract(P); // Fractional part for interpolation
	    vec3 Pf1 = Pf0 - vec3(1.0); // Fractional part - 1.0
	    vec4 ix = vec4(Pi0.x, Pi1.x, Pi0.x, Pi1.x);
	    vec4 iy = vec4(Pi0.yy, Pi1.yy);
	    vec4 iz0 = vec4(Pi0.z);
	    vec4 iz1 = vec4(Pi1.z);
	    
	    vec4 ixy = permute(permute(ix) + iy);
	    vec4 ixy0 = permute(ixy + iz0);
	    vec4 ixy1 = permute(ixy + iz1);
	    
	    vec4 gx0 = ixy0 * (1.0 / 7.0);
	    vec4 gy0 = fract(floor(gx0) * (1.0 / 7.0)) - 0.5;
	    gx0 = fract(gx0);
	    vec4 gz0 = vec4(0.5) - abs(gx0) - abs(gy0);
	    vec4 sz0 = step(gz0, vec4(0.0));
	    gx0 -= sz0 * (step(0.0, gx0) - 0.5);
	    gy0 -= sz0 * (step(0.0, gy0) - 0.5);
	    
	    vec4 gx1 = ixy1 * (1.0 / 7.0);
	    vec4 gy1 = fract(floor(gx1) * (1.0 / 7.0)) - 0.5;
	    gx1 = fract(gx1);
	    vec4 gz1 = vec4(0.5) - abs(gx1) - abs(gy1);
	    vec4 sz1 = step(gz1, vec4(0.0));
	    gx1 -= sz1 * (step(0.0, gx1) - 0.5);
	    gy1 -= sz1 * (step(0.0, gy1) - 0.5);
	    
	    vec3 g000 = vec3(gx0.x,gy0.x,gz0.x);
	    vec3 g100 = vec3(gx0.y,gy0.y,gz0.y);
	    vec3 g010 = vec3(gx0.z,gy0.z,gz0.z);
	    vec3 g110 = vec3(gx0.w,gy0.w,gz0.w);
	    vec3 g001 = vec3(gx1.x,gy1.x,gz1.x);
	    vec3 g101 = vec3(gx1.y,gy1.y,gz1.y);
	    vec3 g011 = vec3(gx1.z,gy1.z,gz1.z);
	    vec3 g111 = vec3(gx1.w,gy1.w,gz1.w);
	    
	    vec4 norm0 = taylorInvSqrt(vec4(dot(g000, g000), dot(g010, g010), dot(g100, g100), dot(g110, g110)));
	    g000 *= norm0.x;
	    g010 *= norm0.y;
	    g100 *= norm0.z;
	    g110 *= norm0.w;
	    vec4 norm1 = taylorInvSqrt(vec4(dot(g001, g001), dot(g011, g011), dot(g101, g101), dot(g111, g111)));
	    g001 *= norm1.x;
	    g011 *= norm1.y;
	    g101 *= norm1.z;
	    g111 *= norm1.w;
	    
	    float n000 = dot(g000, Pf0);
	    float n100 = dot(g100, vec3(Pf1.x, Pf0.yz));
	    float n010 = dot(g010, vec3(Pf0.x, Pf1.y, Pf0.z));
	    float n110 = dot(g110, vec3(Pf1.xy, Pf0.z));
	    float n001 = dot(g001, vec3(Pf0.xy, Pf1.z));
	    float n101 = dot(g101, vec3(Pf1.x, Pf0.y, Pf1.z));
	    float n011 = dot(g011, vec3(Pf0.x, Pf1.yz));
	    float n111 = dot(g111, Pf1);
	    
	    vec3 fade_xyz = fade(Pf0);
	    vec4 n_z = mix(vec4(n000, n100, n010, n110), vec4(n001, n101, n011, n111), fade_xyz.z);
	    vec2 n_yz = mix(n_z.xy, n_z.zw, fade_xyz.y);
	    float n_xyz = mix(n_yz.x, n_yz.y, fade_xyz.x); 
	    return 2.2 * n_xyz;
	}"""
	return code

func _get_code(p_input_vars, p_output_vars):
	var code = ""
	code += p_output_vars[0] + " = cnoise(vec3((UV + " + p_input_vars[0] + ".xy) * " + p_input_vars[1] + ", TIME)) * 0.5 + 0.5;";
	return code

clayjohn reviewed (approved):
I am happy with it once you fix the whitespace error that travis is catching. 

Reduz still needs the final say though.

Chaosus commented:
Added shader type and mode handlers to allow user to override specific behavior.
Also added a documentation (CC @godotengine/documentation). Please check.

bojidar-bg commented:
(slightly offtopic) Would be nice if #12948 is reworked to also directly read `class_name`-s instead of registering node types.

akien-mga commented on doc/classes/VisualShaderNodeCustom.xml:
```diff
@@ -0,0 +1,138 @@
+			<argument index="3" name="type" type="int" enum="VisualShader.Type">
+			</argument>
+			<description>
+				By overriding this function, the returned string will adds the actual shader code. You can modify the output ports values like: [code]return output_vars[0] + " = " + input_vars[0] + ";"[/code]. You can override the generated behavior dependens on the shader [code]mode[/code] (see [constant Shader.Mode]) and/or [code]type[/code] (see [constant VisualShader.Type]).
```
Proofreading:
```
Override this method to define the actual shader code of the associated custom node. The shader code should be returned as a string, which can have multiple lines (the [code]"""[/code] multiline string construct can be used for convenience).
The [code]input_vars[/code] and [code]output_vars[/code] arrays contain the string names of the various input and output variables, as defined by [code]_get_input_*[/code] and [code]_get_output_*[/code] virtual methods in this class.
The output ports can be assigned values in the shader code. For example, [code]return output_vars[0] + " = " + input_vars[0] + ";"[/code].
You can customize the generated code based on the shader [code]mode[/code] (see [enum Shader.Mode]) and/or [code]type[/code] (see [enum VisualShader.Type]).
Defining this method is [b]required[/b].
```

akien-mga commented on doc/classes/VisualShaderNodeCustom.xml:
```diff
@@ -0,0 +1,138 @@
+			<argument index="0" name="mode" type="int" enum="Shader.Mode">
+			</argument>
+			<description>
+				By overriding this function, the returned string will adds the global code on the top of shader. You can define multiple function declarations, varyings, constants, uniforms etc. Be careful with this functionality because it may conflict with other custom nodes (give the defined entities the unique names). You can override the generated behavior dependens on the shader [code]mode[/code] (see  ...
```
Proofreading:
```
Override this method to add shader code on top of the global shader, to define your own standard library of reusable methods, varyings, constants, uniforms, etc. The shader code should be returned as a string, which can have multiple lines (the [code]"""[/code] multiline string construct can be used for convenience).
Be careful with this functionality as it can cause name conflicts with other custom nodes, so be sure to give the defined entities unique names.
You can customize the generated code based on the shader [code]mode[/code] (see [enum Shader.Mode]).
Defining this method is [b]optional[/b].
```

akien-mga commented on doc/classes/VisualShaderNodeCustom.xml:
```diff
@@ -0,0 +1,138 @@
+			<return type="int">
+			</return>
+			<description>
+				By overriding this function, the input port amount is changed to the returned value.
```
Proofreading:
```
Override this method to define the amount of input ports of the associated custom node.
Defining this method is [b]required[/b]. If not overridden, the node has no input ports.
```

akien-mga commented on doc/classes/VisualShaderNodeCustom.xml:
```diff
@@ -0,0 +1,138 @@
+			<return type="int">
+			</return>
+			<description>
+				By overriding this function, the output port amount is changed to the returned value.
```
Proofreading:
```
Override this method to define the amount of output ports of the associated custom node.
Defining this method is [b]required[/b]. If not overridden, the node has no output ports.
```

akien-mga commented on doc/classes/VisualShaderNodeCustom.xml:
```diff
@@ -0,0 +1,138 @@
+			<argument index="0" name="port" type="int">
+			</argument>
+			<description>
+				By overriding this function, the returned string will be used as a name for associated input [code]port[/code].
```
Proofreading:
```
Override this method to define the names of input ports of the associated custom node. The names are used both for the input slots in the editor and as identifiers in the shader code, and are passed in the [code]input_vars[/code] array in [method _get_code].
Defining this method is [b]optional[/b], but recommended. If not overridden, input ports are named as [code]"in" + str(port)[/code].
```

akien-mga commented on doc/classes/VisualShaderNodeCustom.xml:
```diff
@@ -0,0 +1,138 @@
+			<argument index="0" name="port" type="int">
+			</argument>
+			<description>
+				By overriding this function, the returned string will be used as a name for associated output [code]port[/code].
```
Proofreading:
```
Override this method to define the names of output ports of the associated custom node. The names are used both for the output slots in the editor and as identifiers in the shader code, and are passed in the [code]output_vars[/code] array in [method _get_code].
Defining this method is [b]optional[/b], but recommended. If not overridden, output ports are named as [code]"out" + str(port)[/code].
```

akien-mga commented on doc/classes/VisualShaderNodeCustom.xml:
```diff
@@ -0,0 +1,138 @@
+			<argument index="0" name="port" type="int">
+			</argument>
+			<description>
+				By overriding this function, the returned type (see [constant VisualShaderNode.PortType]) will be used for associated input [code]port[/code].
```
Proofreading:
```
Override this method to define the returned type of each input port of the associated custom node (see [enum VisualShaderNode.PortType] for possible types).
Defining this method is [b]optional[/b], but recommended. If not overridden, input ports will return the [constant VisualShaderNode.PORT_TYPE_SCALAR] type.
```

akien-mga commented on doc/classes/VisualShaderNodeCustom.xml:
```diff
@@ -0,0 +1,138 @@
+			<argument index="0" name="port" type="int">
+			</argument>
+			<description>
+				By overriding this function, the returned type (see [constant VisualShaderNode.PortType]) will be used for associated output [code]port[/code].
```
Proofreading:
```
Override this method to define the returned type of each output port of the associated custom node (see [enum VisualShaderNode.PortType] for possible types).
Defining this method is [b]optional[/b], but recommended. If not overridden, output ports will return the [constant VisualShaderNode.PORT_TYPE_SCALAR] type.
```

akien-mga commented on doc/classes/VisualShaderNodeCustom.xml:
```diff
@@ -0,0 +1,138 @@
+			<return type="String">
+			</return>
+			<description>
+				By overriding this function, name of the associated custom node in the members dialog (and graph) will be changed to the returned value.
```
Proofreading:
```
Override this method to define the name of the associated custom node in the Visual Shader Editor's members dialog and graph.
Defining this method is [b]optional[/b], but recommended. If not overridden, the node will be named as "Unnamed".
```

akien-mga commented on doc/classes/VisualShaderNodeCustom.xml:
```diff
@@ -0,0 +1,138 @@
+			<return type="int" enum="VisualShaderNode.PortType">
+			</return>
+			<description>
+				By overriding this function, return icon of the associated custom node in the members dialog will be changed to the returned type.
```
Proofreading:
```
Override this method to define the return icon of the associated custom node in the Visual Shader Editor's members dialog.
Defining this method is [b]optional[/b]. If not overridden, the return icon will be of type [constant VisualShaderNode.PORT_TYPE_SCALAR].
```

Chaosus replied:
Nope, it will be empty

Chaosus replied:
"If not overridden, the return icon will not be placed" - I guess this ok ?

akien-mga replied:
Maybe "If not overridden, no return icon is shown."

Chaosus replied:
Ok, fixed

akien-mga commented on doc/classes/VisualShaderNodeCustom.xml:
```diff
@@ -0,0 +1,138 @@
+			<return type="String">
+			</return>
+			<description>
+			By overriding this function, category of the associated custom node in the members dialog will be changed to the returned value.
```
Indentation is wrong.

Proofreading:
```
Override this method to define the subcategory of the associated custom node in the Visual Shader Editor's members dialog.
Defining this method is [b]optional[/b]. If not overridden, the node will be filed under the root of the main category (see [method _get_category]).
```

akien-mga replied:
I'm not familiar with the VSE yet so it's not clear to me what subcategories are, so I hope I got things right in my reformulation.
May also be worth renaming the method to `_get_subcategory`? I think "subcategory" is more common than "sub category".

Chaosus replied:
Ok..

akien-mga commented on doc/classes/VisualShaderNodeCustom.xml:
```diff
@@ -0,0 +1,138 @@
+	</brief_description>
+	<description>
+		By derivive from this class you can create a custom visual shader script addon which will automatically been added to visual shader editor. By overriding the provided virtual functions you can setup its behavior.
+		In order to correct work you must add a [code]tool[/code] keyword and provide a [code]class_name[/code] for your custom script.
```
We could also make this codeblock a full example of a custom shader. But that might be better suited as an online tutorial referenced below.

akien-mga commented on doc/classes/VisualShaderNodeCustom.xml:
```diff
@@ -0,0 +1,138 @@
+<?xml version="1.0" encoding="UTF-8" ?>
+<class name="VisualShaderNodeCustom" inherits="VisualShaderNode" category="Core" version="3.2">
+	<brief_description>
+		A custom visual shader node.
```
Proofreading:
```
Virtual class to define custom [VisualShaderNode]s for use in the Visual Shader Editor.
```

akien-mga reviewed (commented):
Great work! I did some proofreading of the documentation and left suggestions for improved descriptions for each of them.

Chaosus commented on doc/classes/VisualShaderNodeCustom.xml:
```diff
@@ -0,0 +1,138 @@
+		By derivive from this class you can create a custom visual shader script addon which will automatically been added to visual shader editor. By overriding the provided virtual functions you can setup its behavior.
+		In order to correct work you must add a [code]tool[/code] keyword and provide a [code]class_name[/code] for your custom script.
+	</description>
+	<tutorials>
```
That tutorial is not existed yet, I could create it later and add to this doc ?

akien-mga replied:
Yes of course, I just meant that it would be nice to have in the future.

akien-mga commented:
Thanks!

And awesome work, I'm looking forward to playing with this :D
